fv3core/stencils/dyn_core.py

```diff
@@ -179,381 +197,415 @@ def dyncore_temporaries(shape):
+            # After writing, make 'dp_ref' a K-field and 'zs' an IJ-field
+            state.dp_ref = utils.make_storage_data(
+                state.dp_ref[0, 0, :], (shape[2],), (0,)
             )
+            state.zs = utils.make_storage_data(state.zs[:, :, 0], shape[0:2], (0, 0))
+
+        # "acoustic" loop
+        # called this because its timestep is usually limited by horizontal sound-wave
+        # processes. Note this is often not the limiting factor near the poles, where
+        # the speed of the polar night jets can exceed two-thirds of the speed of sound.
+        for it in range(n_split):
+            # the Lagrangian dynamics have two parts. First we advance the C-grid winds
+            # by half a time step (c_sw). Then the C-grid winds are used to define
+            # advective fluxes to advance the D-grid prognostic fields a full time step
+            # (the rest of the routines).
+            #
+            # Along-surface flux terms (mass, heat, vertical momentum, vorticity,
+            # kinetic energy gradient terms) are evaluated forward-in-time.
+            #
+            # The pressure gradient force and elastic terms are then evaluated
+            # backwards-in-time, to improve stability.
+            remap_step = False
+            if self.namelist.breed_vortex_inline or (it == n_split - 1):
+                remap_step = True
+            if not hydrostatic:
+                if self.do_halo_exchange:
+                    reqs["w_quantity"] = self.comm.start_halo_update(
+                        state.w_quantity, n_points=utils.halo
+                    )
+                if it == 0:
+                    self._set_gz(
+                        state.zs,
+                        state.delz,
+                        state.gz,
+                    )
+                    if self.do_halo_exchange:
+                        reqs["gz_quantity"] = self.comm.start_halo_update(
+                            state.gz_quantity, n_points=utils.halo
+                        )
+            if it == 0:
+                if self.do_halo_exchange:
+                    reqs["delp_quantity"].wait()
+                    reqs["pt_quantity"].wait()
+
+            if it == n_split - 1 and end_step:
+                if self.namelist.use_old_omega:  # apparently True
+                    self._set_pem(
+                        state.delp,
+                        state.pem,
+                        state.ptop,
+                    )
+            if self.do_halo_exchange:
+                reqs_vector.wait()
+                if not hydrostatic:
+                    reqs["w_quantity"].wait()
 
-            riem_solver3.compute(
-                remap_step,
-                dt,
-                akap,
-                state.cappa,
-                state.ptop,
-                state.zs,
-                state.w,
-                state.delz,
-                state.q_con,
+            # compute the c-grid winds at t + 1/2 timestep
```

Comment:
Thought for later: we should be able to communicate this comment by naming c_sw to make it clear it's integrating winds (and only winds) forward on the c-grid, and passing `dt / 2.0` as the timestep.